src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -753,102 +1435,155 @@ public Thread(ThreadGroup group, Runnable target, String name,
+            // zero status corresponds to state "NEW".
+            if (holder.threadStatus != 0)
+                throw new IllegalThreadStateException();
             start0();
-            started = true;
-        } finally {
+        }
+    }
+
+    /**
+     * Schedules this thread to begin execution in the given thread container.
+     * @throws IllegalStateException if the container is shutdown or closed
+     * @throws IllegalThreadStateException if the thread has already been started
+     */
+    void start(ThreadContainer container) {
+        synchronized (this) {
+            // zero status corresponds to state "NEW".
+            if (holder.threadStatus != 0)
+                throw new IllegalThreadStateException();
+
+            boolean started = false;
+            container.onStart(this);  // may throw
             try {
+                // scope locals may be inherited
+                inheritScopeLocalBindings(container);
+
+                // bind thread to container
+                setThreadContainer(container);
+
+                start0();
+                started = true;
+            } finally {
                 if (!started) {
-                    group.threadStartFailed(this);
+                    container.onExit(this);
                 }
-            } catch (Throwable ignore) {
-                /* do nothing. If start0 threw a Throwable then
-                  it will be passed up the call stack */
             }
         }
     }
 
     private native void start0();
 
     /**
-     * If this thread was constructed using a separate
-     * {@code Runnable} run object, then that
-     * {@code Runnable} object's {@code run} method is called;
-     * otherwise, this method does nothing and returns.
-     * <p>
-     * Subclasses of {@code Thread} should override this method.
+     * This method is run by the thread when it executes. Subclasses of {@code
+     * Thread} may override this method.
      *
-     * @see     #start()
-     * @see     #stop()
-     * @see     #Thread(ThreadGroup, Runnable, String)
+     * <p> This method is not intended to be invoked directly. If this thread is a
+     * platform thread created with a {@link Runnable} task then invoking this method
+     * will invoke the task's {@code run} method. If this thread is a virtual thread
+     * then invoking this method directly does nothing.
+     *
+     * @implSpec The default implementation executes the {@link Runnable} task that
+     * the {@code Thread} was created with. If the thread was created without a task
+     * then this method does nothing.
      */
     @Override
     public void run() {
-        if (target != null) {
-            target.run();
+        Runnable task = holder.task;
+        if (task != null) {
+            task.run();
         }
     }
 
     /**
-     * This method is called by the system to give a Thread
+     * Null out reference after Thread termination (JDK-4006245)
+     */
+    void clearReferences() {
```

Comment:
I had a look at the referenced [JDK-4006245](https://bugs.openjdk.java.net/browse/JDK-4006245). The previous/existing reference cleaning logic seems to have been added mainly to `null` the reference to the user/application class instance represented by the `Runnable` task. Of course, other references too were being cleared. The description in the linked JBS seems to suggest that the `Thread` instance wasn't being GCed sooner. That was more than 2 decades back. I don't have enough knowledge of this area to know if this is still an issue in recent GC implementations. 

With the new implementation in this PR, this `clearReferences()` will no longer set the user/application `task` instance to `null`. I understand why we aren't doing that - the `task` field in the `FieldHolder` is `final` (IMO, rightly so). This, in theory, undoes what JDK-4006245 had implemented. I don't know if we still need such a change here and I'm just bringing this up to see if anyone else has any opinions or if any additional experiments need to be run in context of this change.

Reply:
> I had a look at the referenced [JDK-4006245](https://bugs.openjdk.java.net/browse/JDK-4006245). The previous/existing reference cleaning logic seems to have been added mainly to `null` the reference to the user/application class instance represented by the `Runnable` task. Of course, other references too were being cleared. The description in the linked JBS seems to suggest that the `Thread` instance wasn't being GCed sooner. That was more than 2 decades back. I don't have enough knowledge of this area to know if this is still an issue in recent GC implementations.

The comment on that method dates from early JDK releases with the Classic VM and its conservative GC. It also predates the java.lang.ref API.  The question that I think you are asking is if you create a Thread with a runnable task, let it terminate but keep the reference to the Thread forever then what does it keep alive. We can think about more but so far we haven't seen any issues or tests fail with the reference to the runnable task.